### PR TITLE
LibGfx: Add QOI write ability

### PR DESCRIPTION
**Pull request by mhermier**
Hi,

The following change set allows to export files to the QOI file format. When combined with #11642 it also to make exact copies of the reference files from the official site (verified with a test-suite don't think I can publish, since the reference files are quite large).

Some parts need to be polished before merging, but it is fully functional (Maybe it lacks the possibility to override channels like #11642 allow).

### CODE REVIEW

File: AK/Math.h [hunk 1]
```diff
@@ -455,4 +456,20 @@ constexpr T pow(T x, T y)
 
 #undef CONSTEXPR_STATE
 
+// Check that `value` is in `[min,max]`
+template<typename T>
+constexpr bool is_in_bounds(T const& value, T const& min, T const& max)
```
Comment (mads256h) [inline AK/Math.h:461]: I think `is_in_bounds` and `is_in_range` should be `nodiscard`

File: Userland/Libraries/LibGfx/QOILoader.cpp [hunk 2]
```diff
@@ -33,7 +34,7 @@ static ErrorOr<QOIHeader> decode_qoi_header(InputMemoryStream& stream)
     stream >> Bytes { &header, sizeof(header) };
     if (stream.handle_any_error())
         return Error::from_string_literal("Invalid QOI image: end of stream while reading header"sv);
-    if (StringView { header.magic, array_size(header.magic) } != QOI_MAGIC)
+    if (!AK::TypedTransfer<u8>::compare(header.magic, QOI_MAGIC, QOI_MAGIC_SIZE))
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOILoader.cpp:37]: What is being fixed here? The commit message doesn't present any argument for why this change is useful or good.

Also what's up with the `[Ugly]` prefix, please drop that and refrain from trying to add new semantics to commit messages without reaching prior concensus with the community.

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 3]
```diff
@@ -43,4 +45,35 @@ ALWAYS_INLINE static constexpr bool qoi_is_valid_run(u8 run)
     return AK::is_in_bounds(run, QOI_RUN_MIN, QOI_RUN_MAX);
 }
 
+class QOIState {
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOICommon.h:49]: We normally have each class in its own header.

Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOICommon.h:49]: `QOIState` seems overly generic for what this actually does. Please invent a better name?

File: Userland/Libraries/LibGfx/QOILoader.cpp [hunk 4]
```diff
@@ -169,23 +168,21 @@ static ErrorOr<NonnullRefPtr<Bitmap>> decode_qoi_image(InputMemoryStream& stream
                 if (stream.handle_any_error())
                     return Error::from_string_literal("Invalid QOI image: end of stream while reading chunk tag"sv);
                 if (tag == QOI_OP_RGB)
-                    pixel = TRY(decode_qoi_op_rgb(stream, pixel));
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOILoader.cpp:168]: It's not really clear that encapsulating this logic in a class for this single case is a useful abstraction. I personally think it hurts readability and adds unnecessary complexity to the code.

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        if (m_pixels_to_encode == 0)
+            return Error::from_string_literal("Trying to encode using a QOI writer that should be finished instead"sv);
+        --m_pixels_to_encode;
+
+        const i8 dr = pixel.red() - get_previous_pixel().red();
+        const i8 dg = pixel.green() - get_previous_pixel().green();
+        const i8 db = pixel.blue() - get_previous_pixel().blue();
+        const i8 da = pixel.alpha() - get_previous_pixel().alpha();
+
+        false
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:51]: Please no 😁

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+        m_pixels_to_encode = INVALID_PIXELS_TO_ENCODE;
+        TRY(try_encode_qoi_end_marker());
+        return m_stream;
+    }
+
+private:
+    ErrorOr<void> try_encode_qoi_header(u32 width, u32 height, bool has_alpha_channel)
+    {
+        (void)has_alpha_channel;
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:77]: Left over from before it was used?

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+public:
+    ErrorOr<void> start(u32 width, u32 height, bool has_alpha_channel)
+    {
+        if (m_pixels_to_encode != INVALID_PIXELS_TO_ENCODE)
+            return Error::from_string_literal("Trying to start an already initialized QOI writer"sv);
+
+        m_pixels_to_encode = static_cast<u64>(width) * static_cast<u64>(height);
+        VERIFY(m_pixels_to_encode != INVALID_PIXELS_TO_ENCODE);
+
+#if 0
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:28]: Is this dead? Delete?

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+            static_cast<u8>(QOI_OP_RUN | (m_run - 1))
+        };
+        TRY(m_stream.try_append(ReadonlyBytes { bytes }));
+        m_run = 0;
+        return is_equal;
+    }
+
+    ErrorOr<bool> try_encode_qoi_op_index(Color pixel)
+    {
+        u8 index = index_position(pixel);
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:113]: Can index be `auto`?

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+        VERIFY(AK::is_in_bounds<u8>(index, 0, 63));
+        if (get_previously_seen_pixel(index) != pixel)
+            return false;
```
Comment (bgianfo) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:116]: `ErrorOr<bool>` is normally avoided, as it leads to weird semantics and forces you to handle failures twice, like in your quirky little `||` tree above.  Can we make these `ErrorOr<void>`?

File: Userland/Libraries/LibGfx/QOILoader.cpp (hunk 4 above)
Comment (bgianfo) [reply Userland/Libraries/LibGfx/QOILoader.cpp:168]: I now see this is used as a base class in later commits (was reviewing commit by commit). Adding such details to the commit message would valuable to the reviewer.

**Review by bgianfo — CHANGES_REQUESTED** [review]

**Comment by bgianfo** [comment]
@linusg should really take a look as he is the subject matter expert here 😁

File: Userland/Libraries/LibGfx/QOILoader.cpp (hunk 2 above)
Comment (mhermier) [reply Userland/Libraries/LibGfx/QOILoader.cpp:37]: I tagged it as ugly, because I found it ugly, and I need to rethink the solution, or if anyone have a better solution.
The thing is that since in the current code consider the signature as a `StringView` it brings it's problems, which are tried to be resolved by this patch. Even if all the bytes are ASCII compatible a magic are just mere bytes, so using `StringView` here is problematic. This is the reason I used `TypedTransfer` here.

File: Userland/Libraries/LibGfx/QOICommon.h (hunk 3 above)
Comment (mhermier) [reply Userland/Libraries/LibGfx/QOICommon.h:49]: It was done to avoid the proliferation of dedicated headers. There are only a few shared component.

File: Userland/Libraries/LibGfx/QOILoader.cpp (hunk 2 above)
Comment (alimpfard) [reply Userland/Libraries/LibGfx/QOILoader.cpp:37]: `StringView` and `String` don't really promise any specific encoding, but if you're _that_ concerned, why not use `ReadonlyBytes` and its `operator==`?

TypedTransfer is a bit too low-level to use in such a high-level library imo (do note that `Span::operator==` uses this under the hood anyway).

Anyhow, I'm not sure what you find ugly about this (I haven't looked yet), but if you think it's ugly, then why would anyone want to look at it or merge it?

File: Userland/Libraries/LibGfx/QOIWriter.cpp (hunk 5 above)
Comment (mhermier) [reply Userland/Libraries/LibGfx/QOIWriter.cpp:51]: No to `db` `da`? I used the names from the specs, so one can refer to it more easily.
About the `false`, I agree it is a controversial style, but it helped me to quickly disable and trouble shoot parts of the encoder. Now That I have a working base, I can iterate on it.

File: Userland/Libraries/LibGfx/QOIWriter.cpp (hunk 7 above)
Comment (mhermier) [reply Userland/Libraries/LibGfx/QOIWriter.cpp:28]: It is a reminder, that I need to submit a patch to add `ByteBuffer::reserve`

File: Userland/Libraries/LibGfx/QOIWriter.cpp (hunk 6 above)
Comment (mhermier) [reply Userland/Libraries/LibGfx/QOIWriter.cpp:77]: Probably

File: AK/Math.h (hunk 1 above)
Comment (alimpfard) [inline AK/Math.h:461]: These names feel somewhat arbitrary to me, what makes `range` end-exclusive and `bounds` end-inclusive?

(I don't really understand what utility these provide over the imo much more clear boolean expressions they contain, but I don't really care enough to raise that as an actual issue)

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 10]
```diff
@@ -23,7 +23,7 @@ static constexpr u8 QOI_OP_DIFF = 0b01000000;
 static constexpr u8 QOI_OP_LUMA = 0b10000000;
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:27]: The commit title has a typo: `ÈND_MARKER` -> `END_MARKER` (note the E)

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 11]
```diff
@@ -7,6 +7,7 @@
 
 #pragma once
 
+#include <AK/Math.h>
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:11]: Including this rather complicated header just for a bounds check seems silly.

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 12]
```diff
@@ -33,4 +38,9 @@ struct [[gnu::packed]] QOIHeader {
     u8 colorspace;
 };
 
+ALWAYS_INLINE static constexpr bool qoi_is_valid_run(u8 run)
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:44]: I'd say all this stuff has to go in `namespace Gfx::QOI`, `qoi_is_valid_run` is not a very C++ name.

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 13]
```diff
@@ -43,4 +45,35 @@ ALWAYS_INLINE static constexpr bool qoi_is_valid_run(u8 run)
...
+        return (pixel.red() * 3 + pixel.green() * 5 + pixel.blue() * 7 + pixel.alpha() * 11) % PREVIOUSLY_SEEN_PIXELS_SIZE;
+    }
+
+    void set_previous_pixel(Color pixel)
+    {
+        m_previous_pixel = pixel;
+        m_previously_seen_pixels[index_position(pixel)] = pixel;
+    }
+
+    Color get_previous_pixel() const
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:64]: Getters in serenity do not have a `get_` prefix.

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 14]
```diff
@@ -43,4 +45,35 @@ ALWAYS_INLINE static constexpr bool qoi_is_valid_run(u8 run)
     return AK::is_in_bounds(run, QOI_RUN_MIN, QOI_RUN_MAX);
 }
 
+class QOIState {
+public:
+    static constexpr u8 PREVIOUSLY_SEEN_PIXELS_SIZE = 64;
+
+    static inline u8 index_position(Color pixel)
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:53]: This is already implicitly `inline` by being in the class definition in the header file.

File: Userland/Libraries/LibGfx/QOICommon.h [hunk 15]
```diff
@@ -18,7 +18,8 @@ namespace Gfx {
 // Implementation details for the "Quite OK Image" format (v1.0).
 // https://qoiformat.org/qoi-specification.pdf
 
-static constexpr auto QOI_MAGIC = "qoif"sv;
+static constexpr u8 QOI_MAGIC_SIZE = 4;
+static constexpr u8 QOI_MAGIC[QOI_MAGIC_SIZE] = { 'q', 'o', 'i', 'f' };
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOICommon.h:22]: why?

Comment (alimpfard) [reply Userland/Libraries/LibGfx/QOICommon.h:22]: (this could just be a `Span<u8 const>` to make the comparison below even easier)

File: Userland/Libraries/LibGfx/QOIWriter.cpp (hunk 7 above)
Comment (alimpfard) [reply Userland/Libraries/LibGfx/QOIWriter.cpp:28]: You probably want `ByteBuffer::try_ensure_capacity()`.

File: Userland/Libraries/LibGfx/QOIWriter.cpp (hunk 5 above)
Comment (alimpfard) [reply Userland/Libraries/LibGfx/QOIWriter.cpp:51]: No to this cute short-circuit trick.

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+    {
+        (void)has_alpha_channel;
+        QOIHeader header {
+            {},
+            AK::convert_between_host_and_big_endian(width),
+            AK::convert_between_host_and_big_endian(height),
+            static_cast<u8>(has_alpha_channel ? 4 : 3),
+            0,
+        };
+        AK::TypedTransfer<u8>::copy(header.magic, QOI_MAGIC, QOI_MAGIC_SIZE);
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:85]: Please just use `Bytes`.

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        u8 bytes[1] {
+            static_cast<u8>(QOI_OP_INDEX | index)
+        };
+        TRY(m_stream.try_append(ReadonlyBytes { bytes }));
+        return true;
+    }
+
+    ErrorOr<bool> try_encode_qoi_op_diff(i8 dr, i8 dg, i8 db, i8 da)
+    {
+        if (!AK::is_in_bounds<i8>(dr, -2, 1)
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:127]: Instead of explicitly stating the type here, `is_in_bounds`'s signature can only depend on the first argument's type.

File: Userland/Libraries/LibGfx/QOIWriter.cpp [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+ErrorOr<ByteBuffer> QOIWriter::encode(Gfx::Bitmap const& bitmap)
+{
+    const u32 width = static_cast<u32>(bitmap.width());
+    const u32 height = static_cast<u32>(bitmap.height());
+
+    QOIWriterState<> state;
+    TRY(state.start(width, height, bitmap.has_alpha_channel()));
+    for (u32 y = 0; y < height; ++y)
+        for (u32 x = 0; x < width; ++x)
+            TRY(state.encode(bitmap.get_pixel(x, y)));
```
Comment (alimpfard) [inline Userland/Libraries/LibGfx/QOIWriter.cpp:208]: Use braces for blocks with more than a single line.

**Review by alimpfard — CHANGES_REQUESTED** [review]

**Comment by bgianfo** [comment]
@mhermier what's the point of keeping this alive if the comments aren't getting resolved?

**Comment by bgianfo** [comment]
> @mhermier what's the point of keeping this alive if the comments aren't getting resolved?

Asking this again...

**Comment by mhermier** [comment]
Because it can be useful for some, and I'm currently busy with other stuff. So till I'll find some time to resolve the remaining issues, I prefer to keep it alive, than recreating a new PR id every time the bot kills it. This behavior of the bot is good in some reasons, but if one is slow for various reasons, killing PR produce even more noise for nothing...

**Comment by bgianfo** [comment]
> Because it can be useful for some, and I'm currently busy with other stuff. So till I'll find some time to resolve the remaining issues, I prefer to keep it alive, than recreating a new PR id every time the bot kills it.

Please consider just closing the PR and just re-open it when you have time again to work on it in the future. All state will be retained, nothing will be lost.  

Leaving it open endlessly just clogs the PR queue needlessly, and maintainers such as much self end up wasting our time looking at it.

**Comment by mhermier** [comment]
Well, unless I missed the button to reopen the issues, it seems that I'm not able to reopen them (even if I own them). So I don't know if it is a me being blind, a configuration problem on your side or on my side (lack of privilege). I lost a PR ID that I could not reopen by myself, so unless unless that problem of reopening the PR is resolved somehow, I prefer to keep my PR open.
